pkg/statistics/index.go

```diff
@@ -127,23 +127,31 @@ func (idx *Index) TotalRowCount() float64 {
 	return idx.Histogram.TotalRowCount()
 }
 
-// IsInvalid checks if this index is invalid.
-func (idx *Index) IsInvalid(sctx context.PlanContext, collPseudo bool) (res bool) {
-	idx.CheckStats()
+// IndexStatsIsInvalid checks whether the index has valid stats or not.
+func IndexStatsIsInvalid(idxStats *Index, sctx context.PlanContext, coll *HistColl, cid int64) (res bool) {
 	var totalCount float64
 	if sctx.GetSessionVars().StmtCtx.EnableOptimizerDebugTrace {
 		debugtrace.EnterContextCommon(sctx)
 		defer func() {
 			debugtrace.RecordAnyValuesWithNames(sctx,
 				"IsInvalid", res,
-				"CollPseudo", collPseudo,
+				"CollPseudo", coll.Pseudo,
 				"TotalCount", totalCount,
 			)
 			debugtrace.LeaveContextCommon(sctx)
 		}()
 	}
-	totalCount = idx.TotalRowCount()
-	return (collPseudo) || totalCount == 0
+	// If the given index statistics is nil or we found that the index's statistics hasn't been fully loaded, we add this index to NeededItems.
+	// Also, we need to check that this HistColl has its physical ID and it is permitted to trigger the stats loading.
+	if (idxStats == nil || !idxStats.IsFullLoad()) && coll.PhysicalID > 0 && !coll.CanNotTriggerLoad {
```

Comment:
Looks like we can reuse `IndexIsLoadNeeded` here.

Reply:
I think `coll.PhysicalID` won't be equal to or less than 0 and can be removed.

Reply:
The IndexIsLoadNeeded is attached on the `statistics.Table`, not the `statistics.HistColl`